### PR TITLE
feat(723): supports new workflow design

### PR DESCRIPTION
**Pull request by d2lam**
- Support new workflow format
- If there is no `requires` specified, will assume it's an old version
- We should come back and get rid of some of the code once we switch completely to the new format

Blocked by: https://github.com/screwdriver-cd/data-schema/pull/182
Related: https://github.com/screwdriver-cd/screwdriver/issues/723

### CODE REVIEW

File: lib/phase/functional.js [hunk 1]
```diff
@@ -181,14 +181,17 @@ function validateWorkflow(doc) {
 module.exports = flattenedDoc => (
     new Promise((resolve, reject) => {
         const doc = flattenedDoc;
+        const hasRequires = Object.keys(doc.jobs).some(jobName => doc.jobs[jobName].requires);
         let errors = [];
 
         // Jobs
         errors = errors.concat(validateJobSchema(doc));
 
         // Workflow
```
Comment (tkyi) [inline lib/phase/functional.js]: Maybe it'd be helpful to have comments like `legacy workflow` or something for code that we'll need to remove

File: lib/phase/permutation.js [hunk 2]
```diff
@@ -37,6 +37,10 @@ module.exports = validatedDoc => (
                 job.description = doc.jobs[jobName].description;
             }
 
+            if (doc.jobs[jobName].requires !== undefined) {
```
Comment (tkyi) [inline lib/phase/permutation.js]: Probably want to add a comment saying `Check for new config`

Comment (d2lam) [reply lib/phase/permutation.js]: Added comments

File: lib/phase/functional.js (hunk 1 above)
Comment (d2lam) [reply lib/phase/functional.js]: Added some comments

**Comment by jithine** [comment]
@d2lam build failure. otherwise 👍

**Comment by d2lam** [comment]
@jithin1987  Thanks! Build failed because it needs the new data-schema to be merged first :)

**Pull request merged**